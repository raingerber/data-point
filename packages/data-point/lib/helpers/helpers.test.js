/* eslint-env jest */

const _ = require('lodash')
const helpers = require('./helpers')

describe('helpers API', () => {
  test('check what methods are exported in the helpers modules', () => {
    expect(helpers).toMatchSnapshot()
  })
})

describe('helpers.reducify', () => {
  test('pass simple', () => {
    const rpick = helpers.reducify(_.pick)
    const input = ['a', 'c']
    const value = { a: 1, b: 2, c: 3 }
    expect(rpick(input)(value)).toEqual({ a: 1, c: 3 })
  })

  test('pass error', () => {
    const rTest = helpers.reducify(() => {
      throw new Error('testerror')
    })
    expect(() => {
      rTest()({})
    }).toThrow()
  })
})

describe('helpers.reducifyAll', () => {
  test('reducify all keys', () => {
    const rLodash = helpers.reducifyAll(_)
    const input = ['a', 'c']
    const value = { a: 1, b: 2, c: 3 }
    const keys = Object.keys(rLodash)
    expect(keys.length).toBeGreaterThan(10)
    expect(rLodash.pick(input)(value)).toEqual({ a: 1, c: 3 })
  })

  test('reducify specific keys', () => {
    const rLodash = helpers.reducifyAll(_, ['pick', 'map', 'find'])
    const input = ['a', 'c']
    const value = { a: 1, b: 2, c: 3 }
    const keys = Object.keys(rLodash)
    expect(keys).toEqual(['pick', 'map', 'find'])
    expect(rLodash.pick(input)(value)).toEqual({ a: 1, c: 3 })
  })
})

describe('helpers.mockReducer', () => {
  test('test reducerTest', () => {
    const reducerTest = a => (value, acc, done) => {
      done(null, value * a)
    }

    return helpers.mockReducer(reducerTest(2), { value: 100 }).then(result => {
      expect(result.value).toBe(200)
    })
  })
})

describe('helpers.createAccumulator', () => {
  test('It should assign value', () => {
    const acc = helpers.createAccumulator('foo')
    expect(acc).toHaveProperty('value', 'foo')
  })
  test('It should accept options to merge with value', () => {
    const acc = helpers.createAccumulator('foo', {
      context: 'bar'
    })
    expect(acc).toHaveProperty('value', 'foo')
    expect(acc).toHaveProperty('context', 'bar')
  })
})

describe('helpers.createReducerResolver', () => {
  test('It should partially apply dataPoint to method', () => {
    const resolveReducerBound = helpers.createReducerResolver({})
    expect(resolveReducerBound.length).toEqual(2)
  })
})
