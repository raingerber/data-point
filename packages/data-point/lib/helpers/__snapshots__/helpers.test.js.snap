// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers API check what methods are exported in the helpers modules 1`] = `
Object {
  "createAccumulator": [Function],
  "createEntity": [Function],
  "createReducer": [Function],
  "createReducerResolver": [Function],
  "helpers": Object {
    "assign": [Function],
    "constant": [Function],
    "filter": [Function],
    "find": [Function],
    "map": [Function],
    "omit": [Function],
    "parallel": [Function],
    "pick": [Function],
    "withDefault": [Function],
  },
  "isReducer": [Function],
  "mockReducer": [Function],
  "reducify": [Function],
  "reducifyAll": [Function],
  "resolveEntity": [Function],
  "validateEntityModifiers": [Function],
}
`;
